distributed/spans.py

```diff
@@ -36,39 +46,59 @@ def span(*tags: str) -> Iterator[None]:
     >>> d.compute()
 
     In the above example,
-    - Tasks of collections a and b will be annotated on the scheduler and workers with
-      ``{'span': ('my_workflow', 'phase 1')}``
-    - Tasks of collection c (that aren't already part of a or b) will be annotated with
-      ``{'span': ('my_workflow', 'phase 2')}``
-    - Tasks of collection d (that aren't already part of a, b, or c) will *not* be
-      annotated but will nonetheless be attached to span ``('default', )``
+    - Tasks of collections a and b are annotated to belong to span
+      ``('my workflow', 'phase 1')``, 'ids': (<id0>, <id1>)}``;
+    - Tasks of collection c (that aren't already part of a or b) are annotated to belong
+      to span ``('my workflow', 'phase 2')``;
+    - Tasks of collection d (that aren't already part of a, b, or c) are *not*
+      annotated but will nonetheless be attached to span ``('default', )``.
 
     You may also set more than one tag at once; e.g.
     >>> with span("workflow1", "version1"):
     ...     ...
 
+    Finally, you may capture the ID of a span on the client to match it with the
+    :class:`Span` objects the scheduler:
+    >>> cluster = distributed.LocalCluster()
+    >>> client = distributed.Client(cluster)
+    >>> with span("my workflow") as span_id:
+    ...     client.submit(lambda: "Hello world!").result()
+    >>> span = client.cluster.scheduler.extensions["spans"].spans[span_id]
 
-    Note
-    ----
+    Notes
+    -----
     Spans are based on annotations, and just like annotations they can be lost during
-    optimization. Set config ``optimizatione.fuse.active: false`` to prevent this issue.
+    optimization. Set config ``optimization.fuse.active: false`` to prevent this issue.
     """
-    prev_id = dask.config.get("annotations.span", ())
-    with dask.config.set({"annotations.span": prev_id + tags}):
-        yield
+    if not tags:
+        raise ValueError("Must specify at least one span tag")
+
+    prev_tags = dask.config.get("annotations.span.name", ())
+    # You must specify the full history of IDs, not just the parent, because
+    # otherwise you would not be able to uniquely identify grandparents when
+    # they have no tasks of their own.
+    prev_ids = dask.config.get("annotations.span.ids", ())
+    ids = tuple(str(uuid.uuid4()) for _ in tags)
```

Comment:
Why generate a series of IDs here? If I understand correctly, this span (even with multiple tags) will be nested directly under its parent. Is this for simplicity to not have to deal with cases where a single contextmanager declares multiple tags?

Reply:
Yes, this is to simplify the use case of multiple tags declared at once.